packages/wonder-blocks-core/util/unique-id-factory.js

```diff
@@ -0,0 +1,58 @@
+// @flow
+import {IIdentifierFactory} from "../util/types.js";
+
+/**
+ * This is NOT for direct use. Instead, see the UniqueIDProvider component.
+ *
+ * Implements IIdentifierFactory to provide unique identifiers.
+ */
+export default class UniqueIDFactory implements IIdentifierFactory {
+    _uniqueFactoryName: string;
+
+    static _factoryUniquenessCounter = 0;
+
+    /**
+     * Creates a UniqueIDFactory instance.
+     *
+     * @param {string} factoryName An optional case-insensitive name for the
+     * factory. This will be used as part of the identifier. Useful for
+     * providing context to the identifiers, which can be useful in
+     * differentiating elements when debugging the DOM. This must contain only
+     * hyphen and alphanumeric characters.
+     */
+    constructor(scope?: string) {
+        scope = typeof scope === "string" ? scope : "";
+        const normalizedScope = scope.toLowerCase();
+        if (!this._hasValidIdChars(normalizedScope)) {
+            throw new Error(`Invalid factory name: ${scope}`);
+        }
+        this._uniqueFactoryName = `uid-${normalizedScope}-${UniqueIDFactory._factoryUniquenessCounter++}`;
+    }
+
+    _hasValidIdChars(value: ?string) {
+        if (typeof value === "string") {
+            const invalidCharsReplaced = value.replace(/[^\d\w-]/g, "-");
+            if (value === invalidCharsReplaced) {
```

Comment:
Why not
```
return typeof value === "string" ? !`/[^\d\w-]/g.test(value)` : false;
```

Reply:
Legibility?

Reply:
I've reworked this method to be a little cleaner. I think it reads better now.